pennylane/template.py

```diff
@@ -0,0 +1,225 @@
+# pylint: disable=protected-access
+r"""
+QML Templates
+=============
+
+**Module name:** :mod:`pennylane.template`
+
+.. currentmodule:: pennylane.template
+
+This module provides a growing library of templates of common quantum
+machine learning circuit architectures that can be used to easily build
+more complex quantum machine learning models.
+
+For example, you can construct, evaluate, and train quantum circuit based
+on the circuit-centric quantum classifier architecture from
+:cite:`schuld2018circuit` on an arbitrary number of wires and with an
+arbitrary number of blocks in the following way:
+
+.. code-block:: python
+
+    import pennylane as qml
+    from pennylane import numpy as np
+
+    num_wires=4
+    num_blocks=3
+    dev = qml.device('default.qubit', wires=num_wires)
+
+    @qml.qnode(dev)
+    def circuit(weights, x=None):
+        qml.BasisState(x, wires=range(num_wires))
+        qml.template.CircuitCentric(weights, periodic=True, wires=range(num_wires))
+        return qml.expval.PauliZ(0)
+
+    weights=np.random.randn(num_blocks, num_wires, 3)
+    print(circuit(weights, x=np.array(np.random.randint(0, 1, num_wires))))
+
+
+The handy :func:`Interferometer` function can be used to construct arbitrary
+interferometers in terms of elementary :class:`~.Beamsplitter` Operations.
+operations, by giving a lists of beamsplitter parameters. PennyLane can then be used to
+easily differentiate and obviously also optimize these beam splitter angles:
+
+.. code-block:: python
+
+    import pennylane as qml
+    from pennylane import numpy as np
+
+    num_wires=4
+    dev = qml.device('default.gaussian', wires=num_wires)
+    squeezings = np.random.rand(num_wires, 2)
+    num_params = int(num_wires*(num_wires-1)/2)
+    theta = np.random.uniform(0, 2*np.pi, num_params)
+    phi = np.random.uniform(0, 2*np.pi, num_params)
+
+    @qml.qnode(dev)
+    def circuit(theta, phi):
+        for wire in range(num_wires):
+            qml.Squeezing(squeezings[wire][0], squeezings[wire][1], wires=wire)
+
+        qml.template.Interferometer(theta=theta, phi=phi, wires=range(num_wires))
+        return tuple(qml.expval.MeanPhoton(wires=wires) for wires in range(num_wires))
+
+    print(qml.jacobian(circuit, 0)(theta, phi))
+
+The function :func:`CVNeuralNet` implements the continuous variable neural network architecture from :cite:`killoran2018continuous`.
+Provided with a suitable array of weights, such neural networks can now be easily constructed and trained with PennyLane.
+
+Summary
+^^^^^^^
+
+.. autosummary::
+  CircuitCentric
+  CircuitCentricBlock
+  CVNeuralNet
+  CVNeuralNetLayer
+  Interferometer
+
+Code details
+^^^^^^^^^^^^
+"""
+from pennylane.ops import CNOT, Rot, Squeezing, Displacement, Kerr, Beamsplitter
+
+
+def CircuitCentric(weights, periodic=True, ranges=None, imprimitive_gate=CNOT, wires=None):
+    """pennylane.template.CircuitCentric(weights, periodic=True, ranges=None, imprimitive_gate=qml.CNOT, wires)
```

Comment:
Doesn't the signature resolve automatically in sphinx? So no need to duplicate them here?

Reply:
Yes, that's correct. We document the other operations also with just `wires` and not `wires=None` despite `Operation` having a default for wires (actually, why is that?), see here: file:///home/cgogolin/src/pennylane/doc/_build/html/code/ops/cv.html

Reply:
But then why is it optional in the first place?

Reply:
I will leave things as they are and we can then remove the manual signatures, if/when we decide to make `wires` mandatory in `Operation`.